### PR TITLE
Allows you to cmag deep fryers

### PR DESCRIPTION
**Pull request by Contrabang**
## What Does This PR Do
Allows you to cmag deep fryers, causing them to fry themselves, taking 20 seconds per deep fry. However, the more you cmag it, the more likely the chance of it overflowing with some firey foam occurs! If you *somehow* make it to 150 deep fryings (which takes... 50 minutes on an unupgraded deep fryer), you fry your Cmag. Oops. A cmagged deep fryer will still work like normal just... with a longer name. There is a limit on it though, so it doesnt get TOO insane.<br><br>
It also introduces a function allowing for linear burning of colors because... I wanted to burn the color of the deep fryer.

## Why It's Good For The Game
Cmag having more uses is good, and funny. It brings back memories of deep fried super foods and stuff like the [Crashwich](https://wiki.ss13.co/The_Crashwich).

## Images of changes

image
50 minutes later...
image
image

## Testing

Deep fried it a lot.

## Changelog
:cl:
add: Deep fryers can now be cmagged
/:cl:

### CODE REVIEW

File: code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm [hunk 1]
```diff
@@ -84,6 +88,54 @@
...
+		make_foam(deepfried / 3)
+
+/obj/machinery/cooker/deepfryer/checkValid(obj/item/check, mob/user)
+	if(istype(check, /obj/item/card/cmag))
+		if(deepfried >= cmag_limit) // Gotta end the fun sometime...
+			do_sparks(5, 1, src)
+			to_chat(user, "<span class='danger'>You accidently put [check] into [src]!</span>") // DANGER!!! IMPORTANT!!! MY CMAG!!!!
+			return TRUE
+		return FALSE
+	. = ..()
```
Comment (Sirryan2002) [inline code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm:137]:
```suggestion
	return ..()
```

File: code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm [hunk 2]
```diff
@@ -84,6 +88,54 @@
...
+	addtimer(CALLBACK(src, .proc/cmag_finish, user), cooktime)
+
+/obj/machinery/cooker/deepfryer/proc/cmag_finish(mob/user)
+	icon_state = officon
+	soundloop.stop()
+	playsound(loc, 'sound/machines/ding.ogg', 50, 1)
+	on = FALSE
+
+	name = "[thiscooktype] [name]"
+	desc = "[initial(desc)] It has been [thiscooktype] [deepfried] time\s."
```
Comment (Sirryan2002) [inline code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm:120]:
```suggestion
	desc = "[initial(desc)] It has been [thiscooktype] [deepfried] times."
```

File: code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm [hunk 3]
```diff
@@ -84,6 +88,54 @@
...
 		return
 
+/obj/machinery/cooker/deepfryer/cmag_act(mob/user)
+	if(stat & (NOPOWER|BROKEN) || on)
+		return
+	if(panel_open)
+		to_chat(user, "<span class='warning'>Close the panel first!</span>")
+		return
+
+	if(deepfried > cmag_limit) // 1984
```
Comment (Sirryan2002) [inline code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm:98]:
```suggestion
	if(deepfried >= cmag_limit) // 1984
```

File: code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm [hunk 4]
```diff
@@ -84,6 +88,54 @@
...
+
+	ADD_TRAIT(src, TRAIT_CMAGGED, "clown_emag")
+
+	to_chat(user, "<span class='warning'>You deep fry [src]!</span>") // Deep fry the deep fryer
+	deepfried += 1
+
+	icon_state = onicon
+	soundloop.start()
+	on = TRUE
+	addtimer(CALLBACK(src, .proc/cmag_finish, user), cooktime)
```
Comment (Sirryan2002) [inline code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm:111]:
```suggestion
	addtimer(CALLBACK(src, PROC_REF(cmag_finish, user)), cooktime)
```
this needs to use the new `PROC_REF` syntax

File: code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm [hunk 5]
```diff
@@ -84,6 +88,54 @@
...
+
+	name = "[thiscooktype] [name]"
+	desc = "[initial(desc)] It has been [thiscooktype] [deepfried] time\s."
+	var/default_color = color
+	if(isnull(default_color)) // why the fuck is white null? blame lummox
+		default_color = "#FFFFFF"
+	color = color_linear_burn(default_color, foodcolor, 102)
+
+	if(prob(deepfried))
+		add_attack_logs(user, src, "cmagging spilled burning foam")
```
Comment (Sirryan2002) [inline code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm:127]:
```suggestion
		add_attack_logs(user, src, "cmagged a fryer causing burning foam to spill out")
```
this log should be a bit cleaer

File: code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm [hunk 6]
```diff
@@ -84,6 +88,54 @@
...
+
+	if(deepfried > cmag_limit) // 1984
+		return // Handled by checkValid
+
+	. = TRUE
+
+	ADD_TRAIT(src, TRAIT_CMAGGED, "clown_emag")
+
+	to_chat(user, "<span class='warning'>You deep fry [src]!</span>") // Deep fry the deep fryer
+	deepfried += 1
```
Comment (Sirryan2002) [inline code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm:106]:
```suggestion
	deepfried++
```

File: code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm [hunk 7]
```diff
@@ -13,6 +13,10 @@
 	openicon = "fryer_open"
 	has_specials = TRUE
 	upgradeable = TRUE
+	/// How many times have we deep fried ourself?
+	var/deepfried = 0
+	/// Total times it can be cmagged
+	var/cmag_limit = 150
```
Comment (Sirryan2002) [inline code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm:19]: can this just be turned into a define, I see no reason for this to be a variable or be edited

File: code/__HELPERS/icons.dm [hunk 8]
```diff
@@ -627,6 +627,28 @@ world
...
+ * color2 - Color to burn it with
+ * percent - How much we should burn it by (<100 will sometimes lighten, >100 always will darken)
+ *
+ * returns a hex code string
+ */
+/proc/color_linear_burn(color1, color2, percent = 100)
+	var/list/rgb1 = rgb2num(color1)
+	var/list/rgb2 = rgb2num(color2)
+	percent = percent / 100
+	. = "#"
```
Comment (Sirryan2002) [inline code/__HELPERS/icons.dm:643]: the default return should either be null or have a full usable hex color such as `#FFFFFF`

**Comment by Sirryan2002** [comment]
please truncate the name of the fryer to be limited to at max 2 lines. Crashwich stuff is funny but shouldn't be purposefully coded in because this can quickly lag the client if the name is referenced 2-3+ times in quick succession

**Comment by Contrabang** [comment]
Not really motivated on this one anymore

**Pull request closed without merging**